DRILL-8022: Add Provided Schema Support for Excel Reader

cgivre opened the pull request:
# [DRILL-8022](https://issues.apache.org/jira/browse/DRILL-8022): Add Provided Schema Support for Excel Reader

## Description
This PR adds support for provided schema for the Excel format plugin. 

## Documentation
The provided schema functionality is already documented in the Drill docs, but this PR applies it to the Excel reader.

An example query:
```sql
SELECT col1, col2, col3 
FROM table(dfs.`excel/schema_provisioning.xlsx`  
(schema => 'inline=(`col1` INTEGER, `col2` FLOAT, `col3` VARCHAR)'))
```

Previously, the Excel reader only supported dates, and doubles as data types.  This PR adds support for INT data types but only via provided schema.  

One limitation is that while Drill supports nested data, this PR does not allow a user to provision data types that Excel does not support.  So if a user attempts to create a schema with maps, or lists, this will fail. 

## Testing
Added additional unit tests

luocooong commented:
@cgivre Thanks for the contribution. Is it ready for review?

cgivre commented:
> @cgivre Thanks for the contribution. Is it ready for review?

Hi @luocooong This PR is ready for review.  I restarted the CI and it will likely pass this time.

luocooong commented on contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java:
```diff
@@ -363,6 +407,22 @@ private void setFirstRow() {
+   */
+  private void setFirstDataRow() {
+    // Initialize
+    currentRow = rowIterator.next();
```
Is it possible to do this (?) : 
```java
// 1. Remove this line
// 2. Add a loop once
for (int i = 0; i <= rowNumber + 1; i++) {
  currentRow = rowIterator.next();
}
```

cgivre replied:
I suppose so.  Not sure what the advantage is though.

luocooong replied:
The logic has not changed and is easy to understand. Because at least two iterations are required.

luocooong commented on contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java:
```diff
@@ -710,6 +785,40 @@ public void load(Cell cell) {
+      if (cell == null) {
+        columnWriter.setNull();
+      } else {
+        int fieldNumValue = (int)cell.getNumericCellValue();
```
Add a space before `cell`.

cgivre replied:
Fixed

luocooong reviewed (commented):
@cgivre Cool features!
In addition to comments, it is recommended to upgrade the version of `excel-streaming-reader` to the latest (3.1.6). And squash all the commits.

cgivre commented:
@luocooong Thanks for the review.  I addressed your review comments and squashed commits.  Just as an FYI, we changed the settings so that whoever merges the PR can actually do the squash rather than the contributor. 

I should mention that this PR actually has a little bit of an Easter egg... It has a new function in the test utils to make testing functions that produce dates a little easier.

luocooong commented:
@cgivre I have an idea : Is it possible to add a label for the `learning reference`? It can be quick to filter the pull request, and provide a good chance for learning and improvement of the pull request(or called contribution) for all the contributors. Break down the complexity of Drill internal.

cgivre commented:
> @cgivre I have an idea : Is it possible to add a label for the `learning reference`? It can be quick to filter the pull request, and provide a good chance for learning and improvement of the pull request(or called contribution) for all the contributors. Break down the complexity of Drill internal.

Great idea!  Done!

luocooong reviewed (approved):
LGTM. +1
